File: tests/tt_metal/tools/triage/run_operation_chain.py
#!/usr/bin/env python3

# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
# SPDX-License-Identifier: Apache-2.0

import os, glob
import torch
import ttnn
from tests.ttnn.utils_for_testing import assert_with_pcc


def main():
    """Run the operation chain that may hang if infinite loop is present"""

    # Set up environment
    os.environ["TT_METAL_HOME"] = os.getcwd()
    os.environ["PYTHONPATH"] = os.environ["TT_METAL_HOME"]

    # Enable Inspector with RPC server for operation tracking
    os.environ["TT_METAL_INSPECTOR_LOG_PATH"] = "generated/inspector"

    torch.manual_seed(42)

    # Clean up any existing ops.yaml files before test
    ops_files = glob.glob("./generated/inspector/ops/ops.yaml")
    for f in ops_files:
        if os.path.exists(f):
            os.remove(f)

    # Open device
    device = ttnn.open_device(device_id=0)

    try:
        # Create input tensors
        shape = [32, 64]
        torch_a = torch.rand(shape, dtype=torch.bfloat16)
        torch_b = torch.rand(shape, dtype=torch.bfloat16)
        torch_c = torch.rand(shape, dtype=torch.bfloat16)

        # Convert to ttnn tensors
        tensor_a = ttnn.from_torch(torch_a, layout=ttnn.TILE_LAYOUT, device=device)
        tensor_b = ttnn.from_torch(torch_b, layout=ttnn.TILE_LAYOUT, device=device)
        tensor_c = ttnn.from_torch(torch_c, layout=ttnn.TILE_LAYOUT, device=device)

        print("=== Starting operation chain ===")

        # Operation 1: multiply tensor_a by scalar
        print("Step 1: tensor_a * 2.5")
        result1 = ttnn.mul(tensor_a, 2.5)

        # Operation 2: multiply result1 and tensor_b
        print("Step 2: result1 * tensor_b")
        result2 = ttnn.mul(result1, tensor_b)

        # Operation 3: subtract tensor_c from result2
        print("Step 3: result2 - tensor_c")
        result3 = ttnn.subtract(result2, tensor_c)

        # Operation 4: add scalar to result3 (this will hang if infinite loop is present)
        print("Step 4: result3 + 1.0 (may hang if infinite loop is active)")
        final_result = ttnn.add(result3, 1.0)

        print("=== Operation chain complete ===")

        # Convert final result back to torch for verification
        torch_final = ttnn.to_torch(final_result)

        # Compute expected result using torch
        torch_expected = (torch_a * 2.5) * torch_b - torch_c + 1.0

        # Verify the computation is correct
        assert_with_pcc(torch_expected, torch_final, pcc=0.99)

        print("=== Test completed successfully ===")

    finally:
        ttnn.close_device(device)


if __name__ == "__main__":
    main()

File: tests/ttnn/tools/triage/run_operation_chain.cpp
// SPDX-FileCopyrightText: © 2025 Tenstorrent Inc.
//
// SPDX-License-Identifier: Apache-2.0

// This is a standalone C++ version of run_operation_chain.py
// It runs the same operation chain to verify operation tracking and callstack generation
//
// NOTE: This binary is built by CMake (see tests/tt_metal/tools/triage/CMakeLists.txt)
// Output: build/test/tools/triage/run_operation_chain_cpp
//
// The binary is compiled WITHOUT unity build to preserve debug symbols for addr2line.
// This allows proper callstack resolution when debugging operations.

#include <tt-metalium/host_api.hpp>
#include <tt-metalium/mesh_device.hpp>
#include <tt-metalium/tt_metal.hpp>

// Include ttnn headers
#include <ttnn/operations/eltwise/binary/binary.hpp>
#include <ttnn/operations/creation.hpp>
#include <ttnn/operations/functions.hpp>
#include <ttnn/tensor/tensor.hpp>
#include <ttnn/tensor/shape/shape.hpp>

#include <cstdlib>
#include <filesystem>
#include <iostream>
#include <memory>

using namespace tt::tt_metal;

// Step 4: Add scalar to result (this is where the hang will occur in the test)
// Use noinline attribute to ensure this function is not inlined so we can see it in the callstack
tt::tt_metal::Tensor perform_final_add(const tt::tt_metal::Tensor& input) {
    std::cout << "Step 4: result3 + 1.2" << std::endl;
    return ttnn::add(input, 1.2f);
}

// Step 3: Subtract tensor_c and then perform final add
tt::tt_metal::Tensor perform_subtract_and_add(const tt::tt_metal::Tensor& input, const tt::tt_metal::Tensor& tensor_c) {
    std::cout << "Step 3: result2 - tensor_c" << std::endl;
    auto result3 = ttnn::subtract(input, tensor_c);
    return perform_final_add(result3);
}

// Steps 1-2: Multiply operations, then call subtract and add
tt::tt_metal::Tensor perform_operation_chain(
    const tt::tt_metal::Tensor& tensor_a, const tt::tt_metal::Tensor& tensor_b, const tt::tt_metal::Tensor& tensor_c) {
    std::cout << "=== Starting operation chain ===" << std::endl;

    // Operation 1: multiply tensor_a by scalar 2.5 (matching Python: ttnn.mul(tensor_a, 2.5))
    std::cout << "Step 1: tensor_a * 2.5" << std::endl;
    auto result1 = ttnn::multiply(tensor_a, 2.5f);

    // Operation 2: multiply result1 and tensor_b (matching Python: ttnn.mul(result1, tensor_b))
    std::cout << "Step 2: result1 * tensor_b" << std::endl;
    auto result2 = ttnn::multiply(result1, tensor_b);

    // Continue with subtract and add operations
    return perform_subtract_and_add(result2, tensor_c);
}

int main() {
    // Set up environment for inspector
    setenv("TT_METAL_INSPECTOR_LOG_PATH", "generated/inspector", 1);

    // Clean up any existing ops.yaml files before test
    std::filesystem::path ops_file = "./generated/inspector/ops/ops.yaml";
    if (std::filesystem::exists(ops_file)) {
        std::filesystem::remove(ops_file);
    }

    // Create and open device using MeshDevice
    auto device_holder = tt::tt_metal::distributed::MeshDevice::create_unit_mesh(0);
    auto device = device_holder.get();

    try {
        // Create input tensors matching Python test: shape = [32, 64]
        // Note: TTNN uses 4D shapes, so we need [1, 1, 32, 64]
        ttnn::Shape shape({1, 1, 32, 64});

        // Create random tensors (equivalent to torch.rand with seed 42)
        auto tensor_a = ttnn::random::random(shape, DataType::BFLOAT16);
        auto tensor_b = ttnn::random::random(shape, DataType::BFLOAT16);
        auto tensor_c = ttnn::random::random(shape, DataType::BFLOAT16);

        // Convert to device tensors with tile layout
        tensor_a = tensor_a.to_layout(Layout::TILE).to_device(device);
        tensor_b = tensor_b.to_layout(Layout::TILE).to_device(device);
        tensor_c = tensor_c.to_layout(Layout::TILE).to_device(device);

        // Perform the full operation chain
        auto final_result = perform_operation_chain(tensor_a, tensor_b, tensor_c);

        std::cout << "=== Operation chain complete ===" << std::endl;

        // Convert final result back to host for verification
        auto host_result = final_result.cpu().to_layout(Layout::ROW_MAJOR);

        // Basic verification that we got a result with correct shape
        auto result_shape = host_result.logical_shape();
        if (result_shape[0] != 1 || result_shape[1] != 1 || result_shape[2] != 32 || result_shape[3] != 64) {
            std::cerr << "ERROR: Unexpected result shape!" << std::endl;
            device->close();
            return 1;
        }

        std::cout << "=== Test completed successfully ===" << std::endl;

    } catch (const std::exception& e) {
        std::cerr << "Test failed with exception: " << e.what() << std::endl;
        device->close();
        return 1;
    }

    // Close device
    device->close();

    // Note: The actual ops.yaml file will be generated when the Inspector
    // destructor is called at program exit.

    return 0;
}
